protocols/autonatv2/src/request_response.rs

```diff
@@ -0,0 +1,336 @@
+        Request::from_proto(self.next_msg().await?)
+    }
+    pub(crate) async fn next_response(&mut self) -> io::Result<Response> {
+        Response::from_proto(self.next_msg().await?)
+    }
+}
+
+impl<I> Coder<I>
+where
+    I: AsyncWrite + Unpin,
+{
+    async fn send_msg(&mut self, msg: proto::Message) -> io::Result<()> {
+        self.inner.send(msg).await?;
+        Ok(())
+    }
+
+    pub(crate) async fn send_request(&mut self, request: Request) -> io::Result<()> {
+        self.send_msg(request.into_proto()).await
+    }
+
+    pub(crate) async fn send_response(&mut self, response: Response) -> io::Result<()> {
+        self.send_msg(response.into_proto()).await
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum Request {
+    Dial(DialRequest),
+    Data(DialDataResponse),
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct DialRequest {
+    pub(crate) addrs: Vec<Multiaddr>,
+    pub(crate) nonce: u64,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct DialDataResponse {
+    pub(crate) data_count: usize,
+}
+
+impl Request {
+    pub(crate) fn from_proto(msg: proto::Message) -> io::Result<Self> {
+        match msg.msg {
+            proto::mod_Message::OneOfmsg::dialRequest(proto::DialRequest { addrs, nonce }) => {
+                let addrs: Vec<Multiaddr> = addrs
+                    .into_iter()
+                    .map(|e| e.to_vec())
+                    .map(|e| {
+                        Multiaddr::try_from(e).map_err(|err| {
+                            new_io_invalid_data_err!(format!("invalid multiaddr: {}", err))
+                        })
+                    })
+                    .collect::<Result<Vec<_>, io::Error>>()?;
+                let nonce = check_existence!(nonce)?;
+                Ok(Self::Dial(DialRequest { addrs, nonce }))
+            }
+            proto::mod_Message::OneOfmsg::dialDataResponse(proto::DialDataResponse { data }) => {
+                let data_count = check_existence!(data)?.len();
+                Ok(Self::Data(DialDataResponse { data_count }))
+            }
+            _ => Err(new_io_invalid_data_err!(
+                "invalid message type, expected dialRequest or dialDataResponse"
+            )),
+        }
+    }
+
+    pub(crate) fn into_proto(self) -> proto::Message {
+        match self {
+            Request::Dial(DialRequest { addrs, nonce }) => {
+                let addrs = addrs.iter().map(|e| e.to_vec()).collect();
+                let nonce = Some(nonce);
+                proto::Message {
+                    msg: proto::mod_Message::OneOfmsg::dialRequest(proto::DialRequest {
+                        addrs,
+                        nonce,
+                    }),
+                }
+            }
+            Request::Data(DialDataResponse { data_count }) => {
+                debug_assert!(
+                    data_count <= DATA_FIELD_LEN_UPPER_BOUND,
+                    "data_count too large"
+                );
```

Comment:
Can we enforce this upon construction of `DialDataResponse`?

Reply:
I changed it